universalModule/software/RoFIDriver/src/rofi_hal.hpp

```diff
@@ -0,0 +1,344 @@
+     */
+    void setVelocity( float velocity ) { _impl->setVelocity( velocity ); }
+
+    /**
+     * \brief Get current joint position in rad
+     *
+     * This works in all three control modes - position, velocity and torque
+     * mode.
+     */
+    float getPosition() { return _impl->getPosition(); }
+
+    /**
+     * \brief Change the joint's control mode to position and move to position.
+     *
+     * \param pos required position in rad
+     * \param velocity velocity limit in rad/s, required to be positive and
+     * non-zero
+     * \param callback callback to be called once the position is reached
+     */
+    void setPosition( float pos, float velocity,
+        std::function< void( Joint ) > callback )
+    {
+        _impl->setPosition( pos, velocity, callback );
+    }
+
+    /**
+     * \brief Get the current torque of the joint.
+     *
+     * \return torque in N * m
+     */
+    float getTorque() { return _impl->getTorque(); }
+
+    /**
+     * \brief Change the joint's control mode to torque and provide given torque
+     *
+     * \param torque torque setpoint in N * m
+     */
+    void setTorque( float torque ) { return _impl->setTorque( torque ); }
+
+    /**
+     * \brief Set error handling callback
+     */
+    void onError( std::function< void( Joint, Error, const std::string& ) > callback ) {
+        _impl->onError( callback );
+    }
+
+    Joint( std::shared_ptr< Implementation > impl ): _impl( impl ) {}
+private:
+    std::shared_ptr< Implementation > _impl;
+};
+
+enum ConnectorPosition {
+    Retracted = 0,
+    Expanded = 1
+};
+
+enum ConnectorOrientation {
+    North = 0,
+    East = 1,
+    South = 2,
+    West = 3
+};
+
+enum ConnectorLine {
+    Internal = 0,
+    External = 1
+};
+
+/**
+ * \brief Connector state descriptor
+ */
+struct ConnectorState {
+    ConnectorPosition position = ConnectorPosition::Retracted;
+    bool internal = false; ///< Is internal power bus connected to the connector?
+    bool external = false; ///< Is the external power bus connected to the connector?
+    bool connected = false; ///< Is there a mating side connected?
+    ConnectorOrientation orientation = ConnectorOrientation::North;
+    float internalVoltage, internalCurrent;
+    float externalVoltage, externalCurrent;
+};
+
+enum ConnectorEvent {
+    Connected = 0,
+    Disconnected = 1
+};
+
+/**
+ * \brief Proxy for controlling a single connector of RoFI
+ *
+ * Joint cannot be instantiated on its own; you can obtain its instance from
+ * RoFI::getConnector().
+ *
+ * Since the class is only a proxy it can be safely copied and passed around.
+ */
+class Connector {
+public:
+    /**
+     *  \brief Interface of the actual implementation - either physical local,
+     *  physical remote or a simulation one.
+     *
+     *  To provide an implementation, simply inherit from this class.
+     *
+     *  For method description, see description of Connector.
+     */
+    class Implementation {
```

Comment:
Add virtual destructor - or (and possibly more preferable) make the destructor protected